### ISSUE TITLE
Any way to compile to a static site on Netlify?

### ISSUE THREAD

**Issue by tyeeman** [datenstrom/yellow#367]
As the title states, I would like to add yellow to netlify which uses github repository as the source. Netlify copies the source to their servers, compiles to static files, then serves these static files. I have another static cms working there but it is based on GO so the exe just runs to compile. Since Yellow is php I don't know if it's possible to compile without php available on netlify servers? I will look into it a little more.

Update - looks like Netlify does support php commands or I could just upload the local compiled static files to github and let Netlify pull those in. Interesting!

**Comment by markseu** [comment]
I am afraid to say that I am not a big help with Netlify. 

You can build a static website with the following command: `php yellow.php build public` 

This will build a [static website](https://developers.datenstrom.se/help/server-configuration#static-website), complete with everything. According to Netlify's documentation PHP 5.6 and PHP 7.2 is supported, Datenstrom Yellow works with both, so you should be able to configure a continuous deployment from GitHub. I am not sure about the details, perhaps you need to adjust the Netlify settings or create a build script. Please contact [Netlify support](https://www.netlify.com/support/) for more information.

**Comment by tyeeman** [comment]
I'll let you know what they say, thanks.

**Comment by tyeeman** [comment]
Just tried on Netlify and it works pretty well. There is something wrong though with the blog links only. The home and about links work, and the blog main page works, just the individual blog links actually show the html code in the browser, not the proper page decoded by the browser. Take a look - https://lley.netlify.com

**Comment by markseu** [comment]
Yes you're correct, their web server sends back a `Content-Type` header with the wrong MIME type. 

Datenstrom Yellow uses clean URLs and that's why some static files have no `html` extension. A web server should send a response without `Content-Type` header, when the MIME type of a file is unknown, then it will give the web browser a chance to detect the correct type.

See also  [HTTP 1.1](https://tools.ietf.org/html/rfc7231#section-3.1.1.5), [Apache 2.4](http://httpd.apache.org/docs/2.4/mod/core.html#defaulttype), [MIME type for unknown data](https://stackoverflow.com/questions/1176022/unknown-file-type-mime/28652339#28652339)

Here's what you can do:

1. Configure web server to send no default MIME type 
2. Configure web server to send default MIME type `text/html`

Can you please contact Netlify, maybe they can help you with this?

**Comment by tyeeman** [comment]
I have notified Netlify so I'm waiting for their answer. In the meantime is there an edit I can do in the core.php file that adds .html to the end of blog urls? That may be an easy fix, even if it's just for me to use. I don't expect you to modify your code just to make Netlify work but if I know where to edit and make it work it would be much appreciated.

**Comment by tyeeman** [comment]
OK, got it working. Had to add this file to my root github repository - 
netlify.toml
```
[[headers]]
  for = "/blog/*"
  [headers.values]
Content-Type = "text/html"
```

**Comment by markseu** [comment]
Thanks Tony, appreciate your feedback. I think this configuration works for your specific website, but as soon as you put pages into other folders it will not work. It would be good if we could find a generic solution, one that can be copy/pasted ...or Netlify's web server would send no default MIME-type.

**Comment by fool** [comment]
Hi there, I work on the tech support team @ Netlify and took a look into this issue.  In the case of the URL that was quoted:  https://lley.netlify.com - we are actually serving /index.html (as are most web servers, when you request /) so setting a Mime Type of text/html is correct.  Could you give me an example of where we are doing it wrong so I can the reproduction steps to our operations team to analyze?

**Comment by tyeeman** [comment]
@fool , OK, I just removed netlify.toml and now you can see the problem if click into the blog and select any of the three blog posts. Those blog post are served as text/plain but should be text/html.

**Comment by tyeeman** [comment]
@markseu Just FYI I tried this netlify.toml file -

```
[[headers]]
  for = "/*"
  [headers.values]
Content-Type = "text/html"
```
before the one above I posted a couple of days ago. I thought that it would make all site files a mimetype of text/html. It seemed to work but the strange thing was that the css file for all pages could not be read, ie. all pages had no styling. I don't understand that at all, so that's when I added the blog folder in the path.

**Comment by fool** [comment]
That header rule covers literally every file on your site, so not the best workaround.  The one you came up with before was better - covered just the contents of /blog rather than *.css and *.js outside of that directory too.  I've got an issue filed on Netlify's infrastructure about that but no timeline on a fix since this is a pretty rare use case.  We have a feature called "Pretty URL's" that you're intended to use to be able to serve /file.html with a URL of /file in our asset optimization ; I'd suggest you use that rather than publishing html files as /file in the meantime until we get a fix (when I'll follow up in this thread).

**Comment by tyeeman** [comment]
I tried prettyurls feature a couple of days ago but it did nothing because it is meant to do the opposite of this problem. It takes a .html file and serves just the filename without .html.

**Comment by fool** [comment]
yes, you'd have to deploy .html files to use it.

**Comment by markseu** [comment]
Unfortunately, we can't use the "Pretty URL" feature to solve this problem. 

Some static site generators generate files with `html` extension and you need an extra step to rewrite the URL into something pretty. We decided to do this different. Datenstrom Yellow generates files for [clean URLs](https://en.wikipedia.org/wiki/Clean_URL) right away. I am sorry to say, this behaviour can't be changed.

Here's where I think it goes wrong. If the web server sends a HTTP response for a static file, it usually detects the MIME type based on the file extension. There are different MIME types that a web server needs to know, a typical website consists of HTML, CSS, font and image files. If the web server doesn't know the MIME type of a file, for example a file without extension, then there a different ways to proceed. One is to send a HTTP response without `Content-Type` header, another is to send a `Content-Type` header with a default MIME type. The first gives the web browser a chance to detect the correct media type, this is what the Apache web server does. Netlify's web server doesn't do this, it sends a default MIME type for files it doesn't know.

The HTTP protocol says the following in [section 3.1.1.5.](https://tools.ietf.org/html/rfc7231#section-3.1.1.5): 

> A sender that generates a message containing a payload body SHOULD
   generate a Content-Type header field in that message unless the
   intended media type of the enclosed representation is unknown to the
   sender. 

In a nutshell, if the web server doesn't send a "default MIME type" it works out of the box. As a workaround for Netlify, you can try the following. Change the default MIME type to `text/html` or configure `text/html` for locations that don't start with `/media/`. Hope this helps.

**Comment by fool** [comment]
Yes, your bug report was quite clear and I have filed a corresponding bug
on our side.

Since this thread is the only place any of our >300k customers has noticed
the problem, it is a very low priority bug, so I was informing you of the
workaround that would work today.

I have no ETA on the fix for the bug on our side since it is quite
complicated (involving 3 separate services) and since we have many higher
priority issues, so I thought I'd leave you with some suggestion so you
could deploy a site today.

**Comment by markseu** [comment]
Thanks Chris, appreciate your suggestions. 

@tyeeman I would use the first `netlify.toml` file that you suggested. If you put pages into more folders, you can add these folders to your configuration. Here's an example:

```
[[headers]]
  for = "/blog/*"
  [headers.values]
    Content-Type = "text/html; charset=utf-8"

[[headers]]
  for = "/about/*"
  [headers.values]
    Content-Type = "text/html; charset=utf-8"
```

**Comment by tyeeman** [comment]
I just noticed that Yellow is now on staticgen.com with a deploy to netlify button! One of us should try it and see if it actually works. I wonder if they have changed something at their end due to this issue?

**Comment by tyeeman** [comment]
Just loaded Yellow up to me normal web host and I got a similar problem when accessing blog posts. Instead of viewing html code, I got a window asking if I wanted to download the file! So I asked my web host to determine why and he came back with this, which works good - 
```
<Files *>
    ForceType text/html
</Files>
<Files *\.*>
    ForceType None
</Files>
```
So, if netlify has an equivalent to this .htaccess file, I'll be set. I think htaccess files do not work at netlify but I could be wrong.
